backend/app/main.py

```diff
@@ -71,13 +77,54 @@ async def get_current_user(token: str = Depends(oauth2_scheme)):
         raise exc
     except Exception:
         raise HTTPException(status_code=500, detail="An unexpected error occurred.")
-  
+
+async def authenticate_user(email: str, password):
+    """
+    Check if the user exists and verify the password.
+    """
+    user = await users_collection.find_one({"email": email})
+
+    if not user:
+        # print("no user") # debugging purposes
+        return None
+    elif not pbkdf2_sha256.verify(password.get_secret_value(), user["password"]):
+        # print("wrong pass")
+        return None
+    return user
+
 @app.exception_handler(RequestValidationError)
 async def validation_exception_handler(request, exc):
     raise HTTPException(status_code=422, detail="Invalid request")
 
+######################################## LISTINGS ENDPOINTS ########################################
 @app.get(
-    '/listing/{listing_id}',
+    '/search',
+    response_model=SearchGetResponse,
+    responses={'500': {'model': ErrorResponse}},
+)
+async def get_search(
+    query: Optional[str] = Query(None, description="query"), #optional for the sake of pagination
+    limit: Optional[int] = Query(5, description="Number of listings to retrieve", ge=1, le=30), 
+    next: Optional[str] = Query(None, description="Last seen pagination token")
+    ) -> Union[SearchGetResponse, ErrorResponse]:
+    """
+    search listings
+    """
+    try:
+        limit = min(max(limit, 1), 30)
+        try:
+            listings = await get_listings(query=query, limit=limit, next=next)
```

Comment:
Filters not implemented. It's good that you see an opportunity for code de-duplication but I would suggest using a helper method that is shared between the two endpoints instead of modifying the API of the `/listing` endpoint for the sake of `/search`. Moreover, if you want to combine the two endpoints, I also think that might be a good idea. However, whatever solution you settle on, please ensure to update the OpenAPI spec.